### PR TITLE
Only move meta tags to the head when required and add processing for meta[http-equiv]

### LINKED ISSUE #4502: Meta sanitizer erroneously moves HTML5 microdata meta tags to head
## Bug Description

As originally reported in a [support topic](https://wordpress.org/support/topic/error-in-structured-data-4/), the 1.5.x update broke certain Schema.org metadata as output by Yoast.

For example, put this markup in a Custom HTML block:

```html
<span class="td-page-meta" itemprop="author" itemscope itemtype="https://schema.org/Person">
	<meta itemprop="name" content="Siva">
</span>
<meta itemprop="datePublished" content="2020-03-24T18:05:15+05:30">
<meta itemprop="dateModified" content="2020-03-24T18:05:15+05:30">
<meta itemscope itemprop="mainEntityOfPage" itemtype="https://schema.org/WebPage" itemid="https://crictamil.in/mcclenaghan-talks-about-rohith-missing/">
<span class="td-page-meta" itemprop="publisher" itemscope itemtype="https://schema.org/Organization">
	<span class="td-page-meta" itemprop="logo" itemscope itemtype="https://schema.org/ImageObject">
		<meta itemprop="url" content="https://crictamil.in/wp-content/uploads/2018/05/logo.png">
	</span>
	<meta itemprop="name" content="Cric Tamil">
	<meta itemprop="url" content="https://crictamil.in">
</span>
<meta itemprop="headline " content="இவர் மட்டும் இருந்திருந்தா இந்திய அணி நியூஸிலாந்திடம் இவ்வளவு மோசமாக தோற்றிருக்காது - மெக்லனகன் கருத்து">
<span class="td-page-meta" itemprop="image" itemscope itemtype="https://schema.org/ImageObject">
	<meta itemprop="url" content="https://crictamil.in/wp-content/uploads/2020/03/Mcclenaghan-1.jpg">
	<meta itemprop="width" content="1280"><meta itemprop="height" content="720">
</span>
```

When viewing the AMP page, the Custom HTML block then becomes:

```html
<span class="td-page-meta" itemprop="author" itemscope itemtype="https://schema.org/Person"></span>
<span class="td-page-meta" itemprop="publisher" itemscope itemtype="https://schema.org/Organization">
	<span class="td-page-meta" itemprop="logo" itemscope itemtype="https://schema.org/ImageObject"></span>
</span>
<span class="td-page-meta" itemprop="image" itemscope itemtype="https://schema.org/ImageObject"></span>
```

All of the `meta` tags were moved to the head, causing structured data errors.

## Expected Behaviour

The `meta` tags used in the `body` for [HTML5 Microdata](https://developer.mozilla.org/en-US/docs/Web/HTML/Microdata) should not be moved to the `head`.

## Additional context

 - WordPress version: 5.4
 - Plugin version: 1.5.x

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* 

## Implementation brief

The quick fix may be just to omit processing `meta` tags that have the [HTML5 Microdata attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Microdata#Global_attributes):

```diff
--- a/includes/sanitizers/class-amp-meta-sanitizer.php
+++ b/includes/sanitizers/class-amp-meta-sanitizer.php
@@ -70,7 +70,7 @@ class AMP_Meta_Sanitizer extends AMP_Base_Sanitizer {
 	 * Sanitize.
 	 */
 	public function sanitize() {
-		$meta_elements = $this->dom->getElementsByTagName( static::$tag );
+		$meta_elements = $this->dom->xpath->query( '//meta[ not( @itemid ) and not( @itemref ) and not( @itemprop ) and not( @itemscope ) and not( @itemtype ) ]' );
 
 		// Remove all nodes for easy reordering later on.
 		$meta_elements = array_map(
```

There may be other `meta` tags which should be excluded. One possibility is the `property` and `scheme` attributes.

See the generic `meta` tag spec for more: https://github.com/ampproject/amphtml/blob/286a6302fcd007eab303b105c161e76d9e880322/validator/validator-main.protoascii#L701-L727

We may need to rather use a denylist of meta tags rather than an allowlist.

## QA testing instructions

* 

## Demo

* 

## Changelog entry

*

### PR DESCRIPTION
**Pull request by pierlon**
## Summary

Fixes #4502.

* Only move `meta` tags to the `head` when they are required to be moved; allow other tags to be in the `body` (in particular HTML5 Microdata tags).
* Add explicit recognition for `meta[http-equiv]` tags.
* Add tests for all discretely-recognized `meta` tags.
* Add tests to assert the contents of the validator spec to catch unexpected changes.

## Checklist

- [x] My pull request is addressing an [open issue](https://github.com/ampproject/amp-wp/contributing/project-management.md#life-of-an-issue) (please create one otherwise).
- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/contributing/engineering.md#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/contributing/engineering.md) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by westonruter** [comment]
@pierlon Please review my additions.

File: includes/sanitizers/class-amp-meta-sanitizer.php [hunk 1]
```diff
@@ -66,19 +68,41 @@ class AMP_Meta_Sanitizer extends AMP_Base_Sanitizer {
...
+	const BODY_ANCESTOR_META_TAG_SPEC_NAME = 'meta name= and content=';
+
+	/**
+	 * Get tag spec for meta tags which are allowed in the body.
+	 *
+	 * @since 1.5.2
+	 * @return string Deny pattern.
+	 */
+	private function get_body_meta_tag_name_attribute_deny_pattern() {
+		static $pattern = null;
```
Comment (pierlon) [inline includes/sanitizers/class-amp-meta-sanitizer.php:86]: My initial method of directly supplying the pattern was indeed brittle in retrospect. Getting the pattern from the spec and making it static is much more robust.

File: tests/php/test-class-amp-meta-sanitizer.php [hunk 2]
```diff
@@ -30,7 +61,42 @@ public function get_data_for_sanitize() {
 
 		$amp_boilerplate = amp_get_boilerplate_code();
 
-		return [
+		$meta_charset  = '<meta charset="utf-8">';
```
Comment (pierlon) [inline tests/php/test-class-amp-meta-sanitizer.php:64]: These tests here are much more extensive than what I had before and covers the happy path of execution, along with potential mishaps that may occur :+1:.

**Comment by westonruter** [comment]
Pre-release build for testing: [amp.zip](https://github.com/ampproject/amp-wp/files/4422044/amp.zip) (1.5.2-alpha-20200402T153920Z-03af649a3)

**Comment by westonruter** [comment]
Fix verified by user: https://wordpress.org/support/topic/error-in-structured-data-4/page/2/#post-12619449

**Comment by amedina** [comment]
QA: verified meta tags are not being moved to the head.

Screen Shot 2020-04-03 at 12 41 01 PM

Screen Shot 2020-04-03 at 12 40 48 PM

**Pull request merged**